File: 3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.c
int lengthOfLongestSubstring(char* s) {
    int map[128];
    for(int i = 0; i < 128; i++) map[i] = -1;

    int maxLen = 0;
    int start = 0;

    for (int end = 0; s[end] != '\0'; end++){
        char c = s[end];

        if (map[(int)c] >= start){
            start = map[(int)c] + 1;
        }

        map[(int)c] = end;
        int currLen = end - start + 1;
        if (currLen > maxLen){
            maxLen = currLen;
        }
    }

    return maxLen;
}

File: 3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const map = new Map();
    let maxLen = 0
    let start = 0

    for (end = 0; end < s.length; end++){
        const char = s[end]

        if (map.has(char) && map.get(char) >= start){
            start = map.get(char) + 1
        }

        map.set(char, end);
        maxLen = Math.max(maxLen, end - start + 1)
    }

    return maxLen
};